EVG-3209 Expose provider settings in distros REST API

chasinglogic opened the pull request:
I'm not sure if this is should be SuperUser required or not.

john-m-liu commented on rest/model/distro.go:
```diff
@@ -8,8 +8,10 @@ import (
+	Name             APIString               `json:"name"`
+	UserSpawnAllowed bool                    `json:"user_spawn_allowed"`
+	Provider         string                  `json:"provider"`
+	ProviderSettings *map[string]interface{} `json:"provider_settings"`
```
I would make this not a pointer and dereference the field below

chasinglogic replied:
Done

john-m-liu commented on rest/route/distro_routes.go:
```diff
@@ -17,7 +17,7 @@ func getDistroRouteManager(route string, version int) *RouteManager {
 		Methods: []MethodHandler{
 			{
-				Authenticator:  &NoAuthAuthenticator{},
+				Authenticator:  &RequireUserAuthenticator{},
```
Add PrefetchFunctions: []PrefetchFunc{PrefetchUser},

chasinglogic replied:
Done

john-m-liu commented on rest/model/distro.go:
```diff
@@ -18,6 +20,8 @@ func (apiDistro *APIDistro) BuildFromService(h interface{}) error {
 		apiDistro.Name = ToAPIString(v.Id)
 		apiDistro.UserSpawnAllowed = v.SpawnAllowed
+		apiDistro.Provider = v.Provider
+		apiDistro.ProviderSettings = *v.ProviderSettings
```
check for nil before dereferencing ProviderSettings

chasinglogic commented on rest/model/distro.go:
```diff
@@ -8,8 +8,10 @@ import (
+	Name             APIString              `json:"name"`
+	UserSpawnAllowed bool                   `json:"user_spawn_allowed"`
+	Provider         string                 `json:"provider"`
+	ProviderSettings map[string]interface{} `json:"provider_settings,omitempty"`
```
Changed to only give AMI when set omitempty otherwise

chasinglogic commented on rest/model/distro.go:
```diff
@@ -18,6 +20,13 @@ func (apiDistro *APIDistro) BuildFromService(h interface{}) error {
 	case distro.Distro:
 		apiDistro.Name = ToAPIString(v.Id)
 		apiDistro.UserSpawnAllowed = v.SpawnAllowed
+		apiDistro.Provider = v.Provider
```
Done

chasinglogic commented:
I couldn't find an example of going from ProviderSettings to EC2ProviderSettings so I went for the "simple" approach with type assertions. If there is a blessed way to do that let me know and I'm happy to change it.

bsamek commented:
When I `make smoke-start-server` and GET http://localhost:9090/rest/v2/distros, AMI appears as an empty string. It should, however, be omitempty.
```
{
  name: "localhost",
  user_spawn_allowed: false,
  provider: "static",
  ami: ""
}
```

john-m-liu commented on rest/model/distro.go:
```diff
@@ -18,6 +22,20 @@ func (apiDistro *APIDistro) BuildFromService(h interface{}) error {
+		if v.ProviderSettings != nil {
+			ec2Settings := &cloud.EC2ProviderSettings{}
+			err := mapstructure.Decode(v.ProviderSettings, ec2Settings)
+			if err != nil || ec2Settings.AMI == "" {
```
I would handle these 2 scenarios differently - if there is an error return it (it's not too important to manipulate data if something errors because returning an error to the caller means that other return values should not be trusted)

chasinglogic replied:
Changed the parent if condition, now return the error if there is one

tychoish commented on rest/model/distro.go:
```diff
@@ -10,6 +12,8 @@ import (
 	Name             APIString `json:"name"`
 	UserSpawnAllowed bool      `json:"user_spawn_allowed"`
+	Provider         APIString `json:"provider"`
+	AMI              APIString `json:"ami,omitempty"`
```
rename AMI "image_id" and ImageID to be provider agnostic

chasinglogic replied:
Why would we rename it to image ID if as below you only want to do this branch for the EC2 provider? If we're only returning this for the EC2 provider then we should leave this as AMI.

bsamek replied:
The reason for this is for forwards compatibility. We conceivably may wish to add this in the future for some other Cloud provider, e.g. GCE, where image_id might mean something different. Calling this "image_id" will allow us to re-use this API field when we support another provider.

tychoish commented on rest/model/distro.go:
```diff
@@ -18,6 +22,20 @@ func (apiDistro *APIDistro) BuildFromService(h interface{}) error {
 		apiDistro.UserSpawnAllowed = v.SpawnAllowed
+		apiDistro.Provider = ToAPIString(v.Provider)
+
+		if v.ProviderSettings != nil {
```
we must also check if provider == ec2 before attempting this branch

chasinglogic replied:
Done

chasinglogic commented:
Added tests, changed parent if condition to validate that provider is ec2, now return any mapstructure.Decode errors found.

Looks like it still works correctly from my smoke server:

```
{
name: "localhost",
user_spawn_allowed: false,
provider: "static"
}
```

And tests pass:

```
rest/model EVG-3209 λ go test -run TestDistro .
ok  	github.com/evergreen-ci/evergreen/rest/model	(cached)
rest/model EVG-3209 λ
```

Running go test on that package failed but it was a separate test: `TestMetricsSuite` which I'm not sure if it's related or not and didn't look.

tychoish commented:
lgtm, but let's make sure this doesn't break the tests.

john-m-liu commented:
@chasinglogic could you fix the compile issue in the route package, and then we can merge this

richardsamuels commented on rest/route/distro_routes.go:
```diff
@@ -17,9 +17,10 @@ func getDistroRouteManager(route string, version int) *RouteManager {
-				RequestHandler: &distroGetHandler{},
-				MethodType:     http.MethodGet,
+				PrefetchFunctions: []PrefetchFunc{PrefetchUser},
+				Authenticator:     &RequireUserAuthenticator{},
```
@bsamek You are correct. Given that this test is in the smoke test, should we retain the noauth requirement, and censor this information to users without super user permissions?
